cmd/sops/main.go

```diff
@@ -332,3 +348,164 @@ func rotate(c *cli.Context, file string, fileBytes []byte, output io.Writer) err
 	}
 	return nil
 }
+
+func hashFile(filePath string) ([]byte, error) {
+	var result []byte
+	file, err := os.Open(filePath)
+	if err != nil {
+		return result, err
+	}
+	defer file.Close()
+	hash := md5.New()
+	if _, err := io.Copy(hash, file); err != nil {
+		return result, err
+	}
+	return hash.Sum(result), nil
+}
+
+const exampleYaml = `
+# Welcome to SOPS. This is the default template.
+# Remove these lines and add your data.
+# Don't modify the 'sops' section, it contains key material.
+example_key: example_value
+example_array:
+    - example_value1
+    - example_value2
+example_multiline: |
+    this is a
+    multiline
+    entry
+example_number: 1234.5678
+example:
+    nested:
+        values: delete_me
+example_booleans:
+    - true
+    - false
+`
+
+func loadExample(c *cli.Context, file string) (sops.Tree, error) {
+	var in []byte
+	var tree sops.Tree
+	if strings.HasSuffix(file, ".yaml") {
+		in = []byte(exampleYaml)
+	}
+	branch, _ := store(file).Unmarshal(in)
+	tree.Branch = branch
+	ks, err := getKeysources(c, file)
+	if err != nil {
+		return tree, err
+	}
+	tree.Metadata.UnencryptedSuffix = c.String("unencrypted-suffix")
+	tree.Metadata.Version = "2.0.0"
+	tree.Metadata.KeySources = ks
+	key, err := tree.GenerateDataKey()
+	if err != nil {
+		return tree, cli.NewExitError(err.Error(), exitCouldNotRetrieveKey)
+	}
+	tree.Metadata.UpdateMasterKeys(key)
+	return tree, nil
+}
+
+func edit(c *cli.Context, file string, fileBytes []byte) error {
+	var tree sops.Tree
+	var stash map[string][]interface{}
+	var err error
+	if fileBytes == nil {
+		tree, err = loadExample(c, file)
+	} else {
+		tree, stash, err = decryptFile(store(file), fileBytes, c.Bool("ignore-mac"))
+	}
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Could not load file: %s", err), exitCouldNotReadInputFile)
+	}
+	tmpdir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Could not create temporary directory: %s", err), exitCouldNotWriteOutputFile)
+	}
+	defer os.RemoveAll(tmpdir)
+	tmpfile, err := os.Create(path.Join(tmpdir, path.Base(file)))
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Could not create temporary file: %s", err), exitCouldNotWriteOutputFile)
+	}
+	var out []byte
+	if c.Bool("show-master-keys") {
+		out, err = store(file).MarshalWithMetadata(tree.Branch, tree.Metadata)
+	} else {
+		out, err = store(file).Marshal(tree.Branch)
+	}
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Could not marshal tree: %s", err), exitErrorDumpingTree)
+	}
+	_, err = tmpfile.Write(out)
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Could not write output file: %s", err), exitCouldNotWriteOutputFile)
+	}
+	origHash, err := hashFile(tmpfile.Name())
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Could not hash file: %s", err), exitCouldNotReadInputFile)
+	}
+	invalidSyntax := true
+	for invalidSyntax || tree.Metadata.MasterKeyCount() == 0 {
```

Comment:
It looks like invalidSyntax is only false at the very bottom. This variable can be replaced with just `break` at the bottom of the loop. If MasterKeyCount == 0, should this throw some sort of error to the client? It looks like this silently skips over editing the file.